File: express/src/controller/NewsFeedController.ts
import { NextFunction, Request, Response } from 'express';
import { Newsfeed } from '../entity/Newsfeed';
import { iNewsFeedService } from '../service/NewsFeedService';

export class NewsFeedController {
  private service: iNewsFeedService;
  constructor(service: iNewsFeedService) {
    this.service = service;
  }

  all = async (req: Request<{ page: number}, any, any, { size: number, filt: string, ordr: string, dir: boolean}>, res: Response, next: NextFunction) => {
    const { page } = req.params;
    const { size, filt, ordr, dir } = req.query;

    try {
      const newsfeeds = await this.service.findByPageAndSizeAndFilterAndOrder(page, size, filt, ordr, dir);
      res.json({newsfeeds: newsfeeds[0], count: newsfeeds[1]});
    } catch(err) {
        next(err);
    }
  }

  allPersonal = async (req: Request<{ page: number}, any, any, { size: number, filt: string, ordr: string, dir: boolean}>, res: Response, next: NextFunction) => {
    const { page } = req.params;
    const { size, filt, ordr, dir } = req.query;

    try {
      const newsfeeds = await this.service.findByUserAndPageAndSizeAndFilterAndOrder(req.user?.id, page, size, filt, ordr, dir);
      res.json({newsfeeds: newsfeeds[0], count: newsfeeds[1]});
    } catch(err) {
        next(err);
    }
  }

  one = async (req: Request<{id: number}>, res: Response, next: NextFunction) => {
    const { id } = req.params;

    try {
        const newsfeed = await this.service.findOne(id);
        return res.json({ newsfeed });
    } catch(err) {
        next(err);
    }
  }

  save = async (req: Request<any, any, Newsfeed>, res: Response, next: NextFunction) => {
    try {
      const newsfeed = await this.service.save(req.body);
      return res.status(201).json({ newsfeed });
    } catch(err) {
      next(err);
    }
  }

  update = async (req: Request<{id: number}, any, Newsfeed>, res: Response, next: NextFunction) => {
    const { id } = req.params;

    try {
        await this.service.update(id, req.body);
        return res.json({ message: 'Updated succesfully!' });
    } catch(err) {
        next(err);
    }
  }

  remove = async (req: Request<{id: number}>, res: Response, next: NextFunction) => {
    const { id } = req.params;

    try {
        await this.service.remove(id);
        return res.json({ message: 'Deleted succesfully!', id: id });
    } catch(err) {
        next(err);
    }
  }
}

File: spring/src/main/java/hu/marko/szakdolgozat/spring/controller/NewsfeedController.java
package hu.marko.szakdolgozat.spring.controller;

import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.StreamSupport;

import javax.servlet.http.HttpServletRequest;
import javax.validation.constraints.NotNull;

import org.springframework.http.HttpStatus;
import org.springframework.http.ResponseEntity;
import org.springframework.lang.Nullable;
import org.springframework.web.bind.annotation.DeleteMapping;
import org.springframework.web.bind.annotation.GetMapping;
import org.springframework.web.bind.annotation.PathVariable;
import org.springframework.web.bind.annotation.PostMapping;
import org.springframework.web.bind.annotation.PutMapping;
import org.springframework.web.bind.annotation.RequestBody;
import org.springframework.web.bind.annotation.RequestMapping;
import org.springframework.web.bind.annotation.RequestParam;
import org.springframework.web.bind.annotation.RestController;

import hu.marko.szakdolgozat.spring.controller.model.Message;
import hu.marko.szakdolgozat.spring.controller.model.Newsfeed;
import hu.marko.szakdolgozat.spring.controller.model.newModel.NewNewsfeed;
import hu.marko.szakdolgozat.spring.controller.model.wrapper.NewsfeedWrapper;
import hu.marko.szakdolgozat.spring.controller.model.wrapper.NewsfeedsWrapper;
import hu.marko.szakdolgozat.spring.service.NewsfeedService;
import hu.marko.szakdolgozat.spring.service.model.PageModel;
import lombok.AllArgsConstructor;

@AllArgsConstructor
@RestController
@RequestMapping("/api/newsfeeds")
public class NewsfeedController {
  private NewsfeedService newsfeedService;

  @GetMapping("/page/{page}")
  public ResponseEntity<NewsfeedsWrapper> getAllNewsfeeds(@PathVariable("page") @NotNull Integer page,
      @NotNull @RequestParam("size") Integer size,
      @RequestParam("filt") @Nullable String filt, @RequestParam("ordr") @Nullable String ordr,
      @RequestParam("dir") @Nullable Boolean dir) {
    PageModel<hu.marko.szakdolgozat.spring.service.model.Newsfeed> pageModel = newsfeedService
        .findByPageAndSizeAndFilterAndOrder(page, size, filt, ordr, dir);
    List<Newsfeed> newsfeedList = StreamSupport.stream(pageModel.getModels().spliterator(), false).map(Newsfeed::new)
        .collect(Collectors.toList());
    return ResponseEntity.ok().body(new NewsfeedsWrapper(newsfeedList, pageModel.getCount()));
  }

  @GetMapping("/{id}")
  public ResponseEntity<NewsfeedWrapper> getOneNewsfeed(@PathVariable("id") @NotNull Long id) {
    return ResponseEntity.ok().body(new NewsfeedWrapper(new Newsfeed(newsfeedService.findOne(id))));
  }

  @GetMapping("/personal/page/{page}")
  public ResponseEntity<NewsfeedsWrapper> getAllPersonalNewsfeeds(@PathVariable("page") @NotNull Integer page,
      @NotNull @RequestParam("size") Integer size,
      @RequestParam("filt") @Nullable String filt, @RequestParam("ordr") @Nullable String ordr,
      @RequestParam("dir") @Nullable Boolean dir, HttpServletRequest request) {
    Long userId = (Long) request.getAttribute("userId");
    PageModel<hu.marko.szakdolgozat.spring.service.model.Newsfeed> pageModel = newsfeedService
        .findByUserAndPageAndSizeAndFilterAndOrder(userId, page, size, filt, ordr, dir);
    List<Newsfeed> newsfeedList = StreamSupport.stream(pageModel.getModels().spliterator(), false).map(Newsfeed::new)
        .collect(Collectors.toList());
    return ResponseEntity.ok().body(new NewsfeedsWrapper(newsfeedList, pageModel.getCount()));
  }

  @PostMapping("/edit")
  public ResponseEntity<NewsfeedWrapper> saveNewsfeed(@RequestBody @NotNull NewNewsfeed newNewsfeed) {
    return ResponseEntity.status(HttpStatus.CREATED)
        .body(new NewsfeedWrapper(new Newsfeed(newsfeedService.save(newNewsfeed.toServiceNewsfeed()))));
  }

  @PutMapping("/edit/{id}")
  public ResponseEntity<Message> updateNewsfeed(@PathVariable("id") @NotNull Long id,
      @RequestBody @NotNull Newsfeed newsfeed) {
    if (newsfeedService.update(id, newsfeed.toServiceNewsfeed())) {
      return ResponseEntity.ok().body(new Message("Newsfeed update is successful!"));
    }
    return ResponseEntity.badRequest().body(new Message("Couldn't update newsfeed!"));
  }

  @DeleteMapping("/edit/{id}")
  public ResponseEntity<Message> deleteNewsfeed(@PathVariable("id") @NotNull Long id) {
    if (newsfeedService.remove(id) != null) {
      return ResponseEntity.ok().body(new Message("Newsfeed is deleted successfully!"));
    }
    return ResponseEntity.badRequest().body(new Message("Couldn't delete newsfeed!"));
  }
}
